packages/fx-core/src/plugins/resource/cicd/plugin.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { Context } from "@microsoft/teamsfx-api/build/v2";
+import { Inputs, v2, Result, FxError, Platform } from "@microsoft/teamsfx-api";
+import { FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDProviderFactory } from "./providers/factory";
+import { ProviderKind } from "./providers/enums";
+import { questionNames } from "./questions";
+import { InternalError, NoProjectOpenedError } from "./errors";
+import { generateBuildScript } from "./utils/buildScripts";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+export class CICDImpl {
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    // 1. Key inputs (envName, provider, template) x (hostingType, ).
+    if (!inputs.projectPath) {
+      throw new NoProjectOpenedError();
+    }
+    const projectPath = inputs.projectPath;
+    // By default(VSC), get env name from plugin's own `target-env` question.
+    let envName = inputs[questionNames.Environment];
+    if (inputs.platform == Platform.CLI) {
+      // In CLI, get env name from the default `env` question.
+      envName = envInfo.envName;
+    }
+    const providerName = inputs[questionNames.Provider];
+    const templateNames = inputs[questionNames.Template] as string[];
+    if (!envName || !providerName || templateNames.length == 0) {
+      throw new InternalError("Some preconditions of inputs are not met.");
+    }
+
+    telemetryHelper.sendSuccessEvent(context, envInfo, "collect-cicd-questions", {
+      env: envName,
+      provider: providerName,
+      template: templateNames.join(","),
+    });
+
+    // 2. Call factory to get provider instance.
+    const providerInstance = CICDProviderFactory.create(providerName as ProviderKind);
+
+    // 3. Call instance.scaffold(template, replacements: any).
+    //  3.1 Construct replacements.
+    const hostType = context.projectSetting.solutionSettings?.hostType;
+    const replacements = {
+      env_name: envName,
+      build_script: generateBuildScript(context.projectSetting),
+      hosting_type_contains_spfx: hostType == "SPFx",
+      hosting_type_contains_azure: hostType == "Azure",
+      cloud_resources_contains_sql:
+        context.projectSetting.solutionSettings?.["azureResources"].includes("sql") ?? false,
+      api_prefix: context.projectSetting.appName,
+      cloud_resources_contains_apim:
+        context.projectSetting.solutionSettings?.["azureResources"].includes("apim") ?? false,
+    };
+
+    const progressBar = context.userInteraction.createProgressBar(
+      "Scaffolding workflow automation files",
+      templateNames.length
+    );
+
+    await progressBar.start(`Scaffolding workflow file for ${templateNames[0]}.`);
+    await providerInstance.scaffold(projectPath, templateNames[0], replacements);
```

Comment:
I don't quite understand what replacements is. It seems all logic of construct replacement is based on projectSetting, why don't we just pass context or projectSetting to scaffold()?

Reply:
What's the reason for passing templateNames[0]? Why we separate the first one from others and don't include them in parallel or in one for each loop?

Reply:
> I don't quite understand what replacements is. It seems all logic of construct replacement is based on projectSetting, why don't we just pass context or projectSetting to scaffold()?

Replacements are for template rendering (key value pairs - key is placeholder name, value is the target value to be replaced) since there're placeholders in templates.
I prefer not to pass projectSettings directly but put the placements construction out of scaffold:
1. one reason is to separate concerns - template rendering only cares about replacements, and doesn't care where the replacements come from.
2. to provide more flexibility - In the future, we may construct replacements from more than one place (not only from projectSettings).

Reply:
> What's the reason for passing templateNames[0]? Why we separate the first one from others and don't include them in parallel or in one for each loop?

Since progressBar's contract is `1xprogressBar.start + * x progressBar.next`, map the first templateNames[0]  to `progressBar.start`, then all the remaining can be put into one loop.

It's a pretty quick operation, so unnecessary to use parallel. Writing this in sequential order using a loop is also good for the progress bar experience, I mean, compare: 0/1 - 1/1 vs 0/4 - 4/4.